h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoModel.java

```diff
@@ -241,186 +116,16 @@ public String terminate() {
         }
     }
 
-    public static class LeafDecisionPathTracker implements DecisionPathTracker<LeafDecisionPathTracker> {
-        private final AuxInfoLightReader _auxInfo;
-        private boolean _wentRight = false; // Was the last step _right_?
-
-        // OUT
-        private int _nodeId = 0; // Returned when the tree is empty (consistent with SharedTreeNode of an empty tree)
-
-        private LeafDecisionPathTracker(byte[] auxTree) {
-          _auxInfo = new AuxInfoLightReader(new ByteBufferWrapper(auxTree));
-        }
-
-        @Override
-        public boolean go(int depth, boolean right) {
-          if (!_auxInfo.hasNext()) {
-            assert _wentRight || depth == 0; // this can only happen if previous step was _right_ or the tree has no nodes
-            return false;
-          }
-          _auxInfo.readNext();
-          if (right) {
-            if (_wentRight && _nodeId != _auxInfo._nid)
-              return false;
-            _nodeId = _auxInfo.getRightNodeIdAndSkipNode();
-            _auxInfo.skipNodes(_auxInfo._numLeftChildren);
-            _wentRight = true;
-          } else { // left
-            _wentRight = false;
-            if (_auxInfo._numLeftChildren == 0) {
-              _nodeId = _auxInfo.getLeftNodeIdAndSkipNode();
-              return false;
-            } else {
-              _auxInfo.skipNode(); // proceed to next _left_ node
-            }
-          }
-          return true;
-        }
-
-        @Override
-        public LeafDecisionPathTracker terminate() {
-          return this;
-        }
-
-        final int getLeafNodeId() {
-          return _nodeId;
-        }
-    }
-
     public static <T> T getDecisionPath(double leafAssignment, DecisionPathTracker<T> tr) {
-        long l = Double.doubleToRawLongBits(leafAssignment);
-        for (int i = 0; i < 64; ++i) {
-            boolean right = ((l>>i) & 0x1L) == 1;
-            if (! tr.go(i, right)) break;
-        }
-        return tr.terminate();
+      return ScoreTree2.getDecisionPath(leafAssignment, tr);
```

Comment:
This ties the whole think back together. Even though the classes are separate it is still essentially one unit because of the built-in assumptions.

Reply:
This is exactly the place that currently _only_ visualises the otherwise hidden (but still valid) fact, that we support only V2 when we want to call this method. As is stated in the PR description, this PR does not bring any changes. It only rearranges code to be more readable (by properly splitting).

Reply:
Yeah, that seems like a bug to me, https://github.com/h2oai/h2o-3/commit/88e2e2d29168004a8b1e2f46e8c8264a8039c50e - this shouldn't have been limited to 1.2 because LeafNodeAssignment was possible for all the versions. I checked with an older MOJO with this restriction relaxed and it worked with 1.1 as expected.

One of my concern is 

    private static final ScoreTree SCORE_TREE_IMPL_PREFERRED = new ScoreTree2();

I find that misleading because it goes against the concept of looking up the appropriate implementation of `ScoreTree`, storing it in _scoreTree and using it whenever appropriate. Prefered should be the correct version, not the latest.

IMHO there is also no need to delegate the static method getDecisionPath to ScoreTree(anything) because it is very unlikely the representation of the path will ever change. Moving the methods that are meant to be generic to a specific implementation only creates an issue for the future (we will have to move them again to another place).

The issue with the current SharedTreeModel are the `public static` methods that are exposed just for h2o-algos and shouldn't really be public in the first place. The fix could be to extend ScoreTree interface and call the latest implementation from h2o-algos of this interface. Otherwise, we are at risk that someone will just forget to change the reference to SharedTree2 in eg. `computeTreeSubgraph` to the latest version when adding a new implementation of ScoreTree.

I understand your effort and I appreciate it but I don't believe this goes in the right direction.